WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,645 @@
+  comrvHeapEntry_t overlayHeap[];
+} comrvCB_t;
 
 /**
 * local prototypes
 */
+static void  comrvUpdateHeapEntryAccess      (u08_t entryIndex);
+static u08_t comrvGetEvictionCandidates      (u08_t requestedEvictionSize, u08_t* pEvictCandidatesList);
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t token);
 
 /**
 * external prototypes
 */
+extern void  comrv_entry               (void);
+extern void  comrvMemcpyHook           (void* pDest, void* pSrc, u32_t sizeInBytes);
+extern u32_t comrvCrcCalcHook          (void* pAddress, u16_t memprySize);
+extern void  comrvNotificationHook     (u32_t notificationNum, u32_t token);
+extern void* comrvLoadOvlayGroupHook   (u32_t groupOffset, void* pDest, u32_t sizeInBytes);
 
 /**
 * global variables
 */
+comrvStackFrame_t* pStackStartAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_START__;
+comrvCB_t*         pComrvCB;
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
+comrvOverlayToken_t *pOverlayMultiGroupTokensTable = (comrvOverlayToken_t*)&overlayMultiGroupTokensTable;
+
+/**
+* COM-RV initialization function
+*
+* @param  none
+*
+* @return none
+*/
+void comrvInit(void)
+{
+   u08_t              i;
+   void*              pBaseAddress = &__OVERLAY_SEC_START__;
+   comrvStackFrame_t* pStackEndAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_END__ - 1;
+
+   /* determine the location of comrv internal control block */
+   pComrvCB = (comrvCB_t*)pStackStartAddr;// - sizeof(comrvCB_t));
+   /* determine the number of heap entries */
+   pComrvCB->numOfOverlayEntries = D_COMRV_NUM_OF_OVERLAY_ENTRIES;
+   /* determine the location of comrv overlay heap structure */
+   //pComrvCB->overlayHeap = (comrvHeapEntry_t*)((u08_t*)pComrvCB - (pComrvCB->numOfOverlayEntries*sizeof(comrvHeapEntry_t)));
+#ifdef D_COMRV_EVICTION_LRU
+   /* initialize all heap entries */
+   for (i = 0 ; i < pComrvCB->numOfOverlayEntries ; i++)
+   {
+      /* initially each entry points to the previous and next neighbor cells */
+      pComrvCB->overlayHeap[i].prevIndex        = i-1;
+      pComrvCB->overlayHeap[i].nextIndex        = i+1;
+      pComrvCB->overlayHeap[i].token.value      = D_COMRV_ENTRY_TOKEN_INIT_VALUE;
+      pComrvCB->overlayHeap[i].properties.value = D_COMRV_ENTRY_PROPERTIES_INIT_VALUE;
+      pComrvCB->overlayHeap[i].pAddress         = pBaseAddress;
+      pBaseAddress = ((u08_t*)pBaseAddress + D_COMRV_OVL_GROUP_SIZE_MIN);
+   }
+   /* mark the last entry in the LRU list */
+   pComrvCB->overlayHeap[i-1].nextIndex = D_COMRV_LRU_LAST_ITEM;
+   /* set the index of the list LRU and MRU */
+   pComrvCB->lruIndex = 0;
+   pComrvCB->mruIndex = i-1;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+   pStackStartAddr = (comrvStackFrame_t*)&pComrvCB->overlayHeap[pComrvCB->numOfOverlayEntries];
+   /* Init comrv stack - mark the last stack frame */
+   pStackStartAddr->offsetPrevFrame = D_COMRV_END_OF_STACK;
```

Comment:
add todo comment inside the code: make stack check for the comrv stack
add an api "comrv get status" - run till edge of stack

Reply:
done in 0f3602c